src/planet-a/chunk_utils.h

```diff
@@ -251,10 +232,10 @@ class ChunkFillSubdivLUT
 
     void fill_tri_recurse(Vector2us top, Vector2us lft, Vector2us rte, std::uint8_t level);
 
-    //std::unique_ptr<ToSubdiv[]> m_data;
     std::vector<ToSubdiv> m_data;
-    std::uint16_t m_fillVrtxCount{0};
-    std::uint16_t m_edgeVrtxCount{0};
+
+    std::uint16_t m_fillVrtxCount{};
+    std::uint16_t m_edgeVrtxCount{};
```

Comment:
Empty braces like this will zero-initialize (initialize to zero) to prevent an uninitialized value.

I don't have this in guidelines.md, but I'd intend that empty braces means "don't care what this value is, but don't keep it uninitialized," whereas `{0}` would imply that the zero is important.

Reply:
Kind of a philosophical question here, but is there any *actual* difference between "Don't care what the value is, but don't leave it uninitialized" and "don't initialize" ?

Using empty-brace initialization for variables where you don't care what the resulting value is means that automated tools like clang-tidy / msvc-analyzer can't identify "read-from-uninitialized" bugs in the code, which i think is a big risk.

Reply:
Note that I have no problem with the variables here being initialized to zero if that's what's intended. But "initialize them, but i don't care to what" is, in my opinion, worse than 'don't initialize'.

Reply:
And, of course, this whole concept only applies to primative types. Class types / types with constructors implicitly manage their initial state.

Reply:
> Kind of a philosophical question here, but is there any actual difference between "Don't care what the value is, but don't leave it uninitialized" and "don't initialize" ?

If I look at code that looks like this:

```cpp
int value;
do_something(value);
```

`value` is uninitialized and there may be weird inconsistent errors when it's used wrong. static analysis may also scream at it. However, the intention is clear that the initial value of `value` is dont-care.


If I instead do...

```cpp
int value = 0;
do_something(value);
```
there's no risk of using uninitialized values, but the zero looks significant, as if it was a start of a count for example. 


Instead using...
```cpp
int value{};
do_something(value);
```
`value` would be zero-initialized, there's no risk of any weird inconsistent errors. The intention is default or don't-care, nothing is explicitly specified.



> Using empty-brace initialization for variables where you don't care what the resulting value is means that automated tools like clang-tidy / msvc-analyzer can't identify "read-from-uninitialized" bugs in the code, which i think is a big risk.

Are you implying that uninitialized values can help find logic errors?

I believe we should avoid uninitialized values whenever possible. Using uninitialized memory as part of regular calculation logic something I'd mostly consider a bad idea nowadays.

PREfast also tends to scream at every uninitialized value.

Reply:
I don't remember where I originally learned this from. I did find this recent similar discussion through google search results: https://www.reddit.com/r/cpp/comments/1dm2y2t/is_empty_brace_initialization_considered_best/

Since I'm noticing I'm throwing around the term "don't care" quite a bit, I would note that this is a term from digital logic; I'm not sure if it's used in other parts of computing.

Reply:
> Are you implying that uninitialized values can help find logic errors?

Not quite.

I'm saying that from the perspective of a static analyzer

```
int bob;
```
and
```
int alice{};
```

are very different looking.

If the programmer doesn't care about the value, then the programmer doesn't want the value to be read from until something is put there. A static analyzer (e.g. PREfast) can detect this kind of problem with `bob`, but `alice` looks initialized.

My advice is that empty brace initialization for primitive types is a bad idea exactly because of this, and explicitly setting a value is better.

I'd be interested to see an example where PREfast was saying "read from uninitialized!" when that wasn't happening. I suppose it could happen with non-const reference args to functions in other translation units, but that seems like a bug in PREfast that it isn't going and looking at the other translation unit.

Reply:
>> Are you implying that uninitialized values can help find logic errors?
>
> Not quite.

The further explanation makes this a yes? In your explanation, `bob` is kept uninitialized to use it to help find a potential logic error.

So from your point of view, would this example also be unacceptable?
```cpp
int value = 0; // This can be anything, will be overwritten immediately
do_something(value);
```

Since this is more about using uninitialized memory (less about zero-initialization), we're not the first to have this discussion:

https://softwareengineering.stackexchange.com/questions/305886/doesnt-always-initialize-variables-lead-to-important-bugs-being-hidden

I'm mostly on the side of just disallowing uninitialized memory. Undefined behaviour that does stuff like being different between debug vs release is far harder to debug IMO, compared to accidentally reading a zero.

Clang-tidy does have an option for this: `cppcoreguidelines-init-variables`

> I'd be interested to see an example where PREfast was saying "read from uninitialized!" when that wasn't happening....

This alert is from this PR: 

https://github.com/TheOpenSpaceProgram/osp-magnum/security/code-scanning/333

It doesn't even care about whether the variable is being read or not. It just doesn't like uninitialized members.